module/x/peggy/keeper/genesis.go

```diff
@@ -5,12 +5,75 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-func InitGenesis(ctx sdk.Context, keeper Keeper, data types.GenesisState) {
-	keeper.setParams(ctx, data.Params)
+// InitGenesis starts a chain from a genesis state
+func InitGenesis(ctx sdk.Context, k Keeper, data types.GenesisState) {
+	k.SetParams(ctx, *data.Params)
+	// reset valsets in state
+	for _, vs := range data.Valsets {
+		// TODO: block height?
+		k.StoreValsetUnsafe(ctx, vs)
```

Comment:
one should store the validator set per epoch/version (i.e upgrade number) and height number. Eg if you have `peggy-1` as chain-id your epoch is `1`. This is for the case of chains that reset the chain height after upgrades.

Reply:
Lets chat about this in some more detail.